fyrox-impl/src/scene/terrain/brushstroke/mod.rs

```diff
@@ -0,0 +1,967 @@
+    pub fn push(&mut self, message: BrushPixelMessage) {
+        assert!(self.data.len() < self.max_size);
+        if let Some(m) = self
+            .data
+            .iter_mut()
+            .find(|m| m.position == message.position)
+        {
+            if message.alpha > m.alpha {
+                m.alpha = message.alpha;
+                m.value = message.value;
+            }
+        } else {
+            self.data.push_back(message);
+        }
+    }
+}
+
+/// Object to send to painting thread to control which textures are modified.
+#[derive(Debug, Clone)]
+pub struct TerrainTextureData {
+    /// The height and width of the texture in pixels.
+    pub chunk_size: Vector2<u32>,
+    /// The kind of texture.
+    pub kind: TerrainTextureKind,
+    /// The texture resources, organized by chunk grid position.
+    pub resources: FxHashMap<Vector2<i32>, TextureResource>,
+}
+
+/// Terrain textures come in multiple kinds.
+/// Height textures contain f32 values for each vertex of the terrain.
+/// Mask textures contain u8 values indicating transparency.
+/// Coordinates are interpreted differently between the two kinds of texture
+/// because the data in height textures overlaps with the data in neighboring chunks,
+/// so the pixels along each edge are duplicated and must be kept in sync
+/// so that the chunks do not disconnect.
+#[derive(Default, Debug, Copy, Clone, PartialEq, Eq)]
+pub enum TerrainTextureKind {
+    #[default]
+    /// Height texture with f32 height values and overlapping edges between chunks.
+    Height,
+    /// Mask texture with u8 oppacity values.
+    Mask,
+}
+
+/// Sender with methods for sending the messages which control a brush painting thread.
+pub struct BrushSender(Sender<BrushThreadMessage>);
+
+impl BrushSender {
+    /// Create a new BrushSender using the given Sender.
+    pub fn new(sender: Sender<BrushThreadMessage>) -> Self {
+        Self(sender)
+    }
+    /// Begin a new stroke using the given brush.
+    pub fn start_stroke(&self, brush: Brush, node: Handle<Node>, data: TerrainTextureData) {
+        self.0
+            .send(BrushThreadMessage::StartStroke(brush, node, data))
+            .unwrap_or_else(on_send_failure);
+    }
+    /// End the current stroke.
+    pub fn end_stroke(&self) {
+        self.0
+            .send(BrushThreadMessage::EndStroke)
+            .unwrap_or_else(on_send_failure);
+    }
+    /// Draw a pixel using the brush that was set in the most recent call to [BrushSender::start_stroke].
+    #[inline]
+    pub fn draw_pixel(&self, position: Vector2<i32>, alpha: f32, value: f32) {
+        if alpha == 0.0 {
+            return;
+        }
+        self.0
+            .send(BrushThreadMessage::Pixel(BrushPixelMessage {
+                position,
+                alpha,
+                value,
+            }))
+            .unwrap_or_else(on_send_failure);
+    }
+}
+
+fn on_send_failure(_error: SendError<BrushThreadMessage>) {
+    /*

```

Comment:
Perhaps that should be uncommented. It seems right to log an error when the channel fails to send, but the reason I commented it was because when this error happens it means that the background thread has panicked and spamming the log with fail-to-send messages makes it harder to find the actually important message that explains why the thread panicked.